### PR TITLE
Cherry pick all 0.4.x

### PR DESCRIPTION
**Pull request by hicqu**
task list:
- [x] cherry pick all commits from 0.4.x to master
- [x] fix all test cases

### CODE REVIEW

**Comment by hicqu** [comment]
@nrc some commits are empty because the work has been done in both master and 0.4.x branch. Let's keep the original commit log so that we can trace it one by one.
Now all test cases are fixed. PTAL thank you!

**Comment by BusyJay** [comment]
@overvenus Would you like to take a look?

File: src/raft.rs [hunk 1]
```diff
@@ -2062,16 +2066,55 @@ impl<T: Storage> Raft<T> {
...
+            return;
+        }
         if m.index < self.raft_log.committed {
             debug!(
                 self.logger,
                 "got message with lower index than committed.";
                 "tag" => &self.tag,
             );
             let mut to_send = Message::default();
+            to_send.to = m.from;
```
Comment (overvenus) [inline src/raft.rs]: Dup with line 2119.

File: src/raft.rs [hunk 2]
```diff
@@ -2115,7 +2158,12 @@ impl<T: Storage> Raft<T> {
     /// For a message, commit and send out heartbeat.
     pub fn handle_heartbeat(&mut self, mut m: Message) {
         self.raft_log.commit_to(m.commit);
+        if self.pending_request_snapshot != INVALID_INDEX {
+            self.send_request_snapshot();
+            return;
+        }
         let mut to_send = Message::default();
+        to_send.to = m.from;
```
Comment (overvenus) [inline src/raft.rs]: Same as above.

File: src/storage.rs [hunk 3]
```diff
@@ -669,18 +689,23 @@ mod test {
...
-        for (i, (idx, wresult)) in tests.drain(..).enumerate() {
+        for (i, (idx, wresult, windex)) in tests.drain(..).enumerate() {
             let storage = MemStorage::new();
             storage.wl().entries = ents.clone();
             storage.wl().raft_state.hard_state.commit = idx;
             storage.wl().raft_state.hard_state.term = idx;
             storage.wl().raft_state.conf_state = conf_state.clone();
 
-            let result = storage.snapshot();
+            let result = storage.snapshot(windex);
```
Comment (overvenus) [inline src/storage.rs]: We should trigger unavilable if `wresult` is an error.

File: tests/integration_cases/test_raft.rs [hunk 4]
```diff
@@ -4364,3 +4291,284 @@ fn test_conf_change_check_before_campaign() {
     }
     assert_eq!(nt.peers[&1].state, StateRole::Candidate);
 }
+
+fn prepare_request_snapshot() -> (Network, Snapshot) {
+    /************************
```
Comment (overvenus) [inline tests/integration_cases/test_raft.rs]: Please uncomment or clean up.

File: src/storage.rs (hunk 3 above)
Comment (hicqu) [reply src/storage.rs]: Could you explain it with more details? I thinks it's already what we want.

File: tests/integration_cases/test_raft.rs (hunk 4 above)
Comment (hicqu) [reply tests/integration_cases/test_raft.rs]: I guess you review the code before the `fix test` commit.

Comment (overvenus) [reply tests/integration_cases/test_raft.rs]: Oh, I missed it.

**Comment by BusyJay** [comment]
This is blocked by #264.

File: src/raft.rs [hunk 5]
```diff
@@ -594,14 +595,14 @@ impl<T: Storage> Raft<T> {
...
         let mut is_batched = false;
         for msg in &mut self.msgs {
-            if msg.msg_type() == MessageType::MsgAppend && msg.to == to {
+            if msg.get_msg_type() == MessageType::MsgAppend && msg.to == to {
                 if !ents.is_empty() {
                     if !util::is_continuous_ents(msg, ents) {
                         return is_batched;
                     }
-                    let mut batched_entries = msg.take_entries();
+                    let mut batched_entries: Vec<_> = msg.take_entries().into();
```
[the remark this answers is not in the document]
Comment (BusyJay) [reply src/raft.rs]: `take_entries()` return a `RepeatedField`.

[the remark this answers is not in the document]
Comment (overvenus) [reply src/raft.rs]: `RepeatedField` has `push` too, so I think the conversion is unnecessary.

[the remark this answers is not in the document]
Comment (BusyJay) [reply src/raft.rs]: The conversion is to support switching prost and rust-protobuf without modification of code.

**Pull request merged**